src/Sonic/Utils.hs

```diff
@@ -26,33 +28,24 @@ hadamardp a b | length a == length b = zipWith (*) a b
 dot :: Num a => [a] -> [a] -> a
 dot xs ys = sum $ hadamardp xs ys
 
-(^+^) :: Num a => [a] -> [a] -> [a]
-(^+^) = zipWith (+)
-
-(^-^) :: Num a => [a] -> [a] -> [a]
-(^-^) = zipWith (-)
-
-evalOnY :: (Num f, Eq f, Fractional f) => f -> Laurent (Laurent f) -> Laurent f
+evalOnY :: Fr -> BiVariateLaurent Fr -> Laurent Fr
 evalOnY y l
   = newLaurent (expLaurent l) ((\l' -> evalLaurent l' y) <$> (coeffsLaurent l))
 
-evalOnX :: (Num f, Eq f, Fractional f) => f -> Laurent (Laurent f) -> Laurent f
+evalOnX :: Fr -> BiVariateLaurent Fr -> Laurent Fr
 evalOnX x l
   = sum $ zipWith f [expLaurent l ..] (coeffsLaurent l)
   where
     f ex lau
       | ex >= 0 = lau `multLaurent` (newLaurent 0 [x ^ ex])
       | otherwise = lau `multLaurent` (newLaurent 0 [recip x ^ (abs ex)])
```

Comment:
Use `pow` instead of (^) always - this way you can do `pow x ex` regardless of the sign of `ex`